File: 08-K Means/ex8.m
fprintf('Finding closest centroids.\n\n');
load('ex7data2.mat');

function idx = findClosestCentroids(X, centroids) 
  K = size(centroids, 1);
  idx = zeros(size(X,1), 1);
  for i = 1:length(idx)
	  mindist = realmax();
	  for k = 1:K
		  d = norm(X(i,:) - centroids(k,:));
		  if (d < mindist)
			  mindist = d;
			  idx(i) = k;
		  endif
	  endfor
  endfor  
endfunction

function centroids = computeCentroids(X, idx, K)
  [m n] = size(X);
  centroids = zeros(K, n);
  for k = 1:K
	  idx_k = find(idx==k);
	  centroids(k,:) = sum(X(idx_k,:),1) / length(idx_k);
  endfor
endfunction

function [centroids, idx] = runkMeans(X, initial_centroids, max_iters, plot_progress)
  if ~exist('plot_progress', 'var') || isempty(plot_progress)
    plot_progress = false;
  endif  
  if plot_progress
    figure;
    hold on;
  endif
  
  [m n] = size(X);
  K = size(initial_centroids, 1);
  centroids = initial_centroids;
  previous_centroids = centroids;
  idx = zeros(m, 1);

  for i=1:max_iters
    fprintf('K-Means iteration %d/%d...\n', i, max_iters);
    if exist('OCTAVE_VERSION')
       fflush(stdout);
    endif
    idx = findClosestCentroids(X, centroids);
    if plot_progress
       plotProgresskMeans(X, centroids, previous_centroids, idx, K, i);
       previous_centroids = centroids;
       fprintf('Press enter to continue.\n');
       pause;
    endif
    centroids = computeCentroids(X, idx, K);
  endfor
  if plot_progress
    hold off;
  endif
endfunction

function centroids = kMeansInitCentroids(X, K)
  centroids = zeros(K, size(X, 2));
  randidx = randperm(size(X, 1));
  centroids = X(randidx(1:K), :);
endfunction

K = 3;
initial_centroids = [3 3; 6 2; 8 5];
idx = findClosestCentroids(X, initial_centroids);
fprintf('Closest centroids for the first 3 examples: \n')
fprintf(' %d', idx(1:3));
fprintf('\n');

fprintf('Program paused. Press enter to continue.\n');
pause;

fprintf('\nComputing centroids means.\n\n');
centroids = computeCentroids(X, idx, K);
fprintf('Centroids computed after initial finding of closest centroids: \n')
fprintf(' %f %f \n' , centroids');
fprintf('Program paused. Press enter to continue.\n');
pause;

fprintf('\nRunning K-Means clustering on example dataset.\n\n');
load('ex7data2.mat');
K = 3;
max_iters = 10;
initial_centroids = [3 3; 6 2; 8 5];
[centroids, idx] = runkMeans(X, initial_centroids, max_iters, false);
fprintf('\nK-Means Done.\n\n');

fprintf('\nRunning K-Means clustering on pixels from an image.\n\n');
A = double(imread('bird_small.png'));
A = A / 255;
img_size = size(A);
X = reshape(A, img_size(1) * img_size(2), 3);
K = 16; 
max_iters = 10;
initial_centroids = kMeansInitCentroids(X, K);
[centroids, idx] = runkMeans(X, initial_centroids, max_iters, false);

fprintf('Program paused. Press enter to continue.\n');
pause;

fprintf('\nApplying K-Means to compress an image.\n\n');
idx = findClosestCentroids(X, centroids);
X_recovered = centroids(idx,:);
X_recovered = reshape(X_recovered, img_size(1), img_size(2), 3);

subplot(1, 2, 1);
imagesc(A); 
title('Original');

subplot(1, 2, 2);
imagesc(X_recovered)
title(sprintf('Compressed, with %d colors.', K));

fprintf('Program paused. Press enter to continue.\n');
pause;

sel = floor(rand(1000, 1) * size(X, 1)) + 1;
palette = hsv(K);
colors = palette(idx(sel), :);
figure;
scatter3(X(sel, 1), X(sel, 2), X(sel, 3), 10, colors);
title('Pixel dataset plotted in 3D. Color shows centroid memberships');
fprintf('Program paused. Press enter to continue.\n');
pause;

function [X_norm, mu, sigma] = featureNormalize(X)
  mu = mean(X);
  X_norm = bsxfun(@minus, X, mu);
  sigma = std(X_norm);
  X_norm = bsxfun(@rdivide, X_norm, sigma);
endfunction

function [U, S] = pca(X)
  [m, n] = size(X);
  U = zeros(n);
  S = zeros(n);
  Sigma = (X' * X) / m; 
  [U, S, V] = svd(Sigma);
endfunction

function Z = projectData(X, U, K)
  Z = zeros(size(X, 1), K);
  for i = 1:size(Z,1)
	  x = X(i, :)';
	  Z(i,:) = x' * U(:, 1:K);
  endfor
endfunction

figure;
Z = projectData(X_norm, U, 2);
plotDataPoints(Z(sel, :), idx(sel), K);
title('Pixel dataset plotted in 2D, using PCA for dimensionality reduction');
fprintf('Program paused. Press enter to continue.\n');
pause;
[X_norm, mu, sigma] = featureNormalize(X);
[U, S] = pca(X_norm);
#Z = projectData(X_norm, U, 2);

File: 08-K Means/ex8.py
import numpy as np
import scipy.io as io
import matplotlib.pyplot as plt
import cv2

def k_means_init_centroids(X, K):
    rand_idx = np.random.permutation(X.shape[0])
    centroids = X[rand_idx[0:K], :]
    return centroids

def find_closest_centroids(X, centroids):
    m = X.shape[0]
    idx = np.zeros(m)
    for i in range(m):
        dist = np.sum(np.square(centroids - X[i, :]), axis=1)
        idx[i] = np.argmin(dist)
    return idx

def compute_centroids(X, idx, K):
    m, n = X.shape
    centroids = np.zeros((K, n))
    for k in range(K):
        x = X[idx == k]
        centroids[k, :] = np.mean(x, axis=0)
    return centroids

def run_k_means(X, initial_centroids, max_iters):
    K = initial_centroids.shape[0]
    centroids = initial_centroids
    history_centroids = np.zeros((max_iters, centroids.shape[0], centroids.shape[1]))
    idx = np.zeros(X.shape[0])

    for i in range(max_iters):
        print ('K-Means iteration {}/{}'.format(i + 1, max_iters))
        history_centroids[i, :] = centroids
        idx = find_closest_centroids(X, centroids)
        centroids = compute_centroids(X, idx, K)
    return centroids, idx

if __name__ == "__main__":

    # ==================== Part 1: Find Closest Centroids ====================
    print ('Finding closest centroids...')
    data = io.loadmat('ex7data2.mat')
    X = data['X']
    K = 3
    initial_centroids = k_means_init_centroids(X,K)
    idx = find_closest_centroids(X, initial_centroids)

    # ===================== Part 2: Compute Means =========================
    print ('Computing centroids means...')
    centroids = compute_centroids(X, idx, K)
    print(centroids)

    # =================== Part 3: K-Means Clustering ======================
    print ('Running K-Means clustering on example dataset...')
    max_iters = 10
    centroids, idx = run_k_means(X, initial_centroids, max_iters)
    print ('K-Means Done.')

    # ============= Part 4: K-Means Clustering on Pixels ===============
    print ('Running K-Means clustering on pixels from an image.')
    A = cv2.imread('bird_small.png')
    A = A.astype(float)/255
    img_size = A.shape

    # Reshape the image into an Nx3 matrix where N = number of pixels.
    X = A.reshape([img_size[0] * img_size[1], img_size[2]])
    K = 16
    initial_centroids = k_means_init_centroids(X, K)
    centroids, idx = run_k_means(X, initial_centroids, max_iters)

    # ================= Part 5: Image Compression ======================
    print ('Applying K-Means to compress an image.')
    idx = find_closest_centroids(X, centroids)

    # Recover the image from the indices (idx) by mapping each pixel to the centroid value.
    X_recovered = centroids[idx.astype(int), :]

    # Reshape the recovered image into proper dimensions
    X_recovered = X_recovered.reshape(img_size)

    fig = plt.figure()
    # Display the original image
    ax1 = fig.add_subplot(1, 2, 1)
    ax1.imshow(A)
    ax1.set_title('Original')
    # Display compressed image side by side
    ax2 = fig.add_subplot(1, 2, 2)
    ax2.imshow(X_recovered)
    ax2.set_title('Compressed, with {} colors.'.format(K))
    plt.show()